lib/active_merchant/billing/gateways/square.rb

```diff
@@ -0,0 +1,375 @@
+      # The `test` indicates if these credentials are for sandbox or production (money moving) access
+      def initialize(options={})
+        requires!(options, :login, :password, :location_id, :test)
+        @client_id = options[:login].strip
+        @bearer_token = options[:password].strip
+        @location_id = options[:location_id].strip
+
+        super
+      end
+
+      # To create a charge on a card using a card nonce:
+      #     purchase(money, card_nonce, { ...create transaction options... })
+      #
+      # To create a customer and save a card (via card_nonce) to the customer:
+      #     purchase(money, card_nonce, {customer: {...params hash same as in store() method...}, ...})
+      #   Note for US and CA, you must have {customer: {billing_address: {zip: xxxxx}}} which passes
+      #   AVS to store a card.
+      #   Note this always creates a new customer, so it may make a duplicate customer if this card was
+      #   associated to another customer previously.
+      #
+      # To use a customer's card on file:
+      #     purchase(money, nil, {customer: {id: 'customer-id', card_id: 'card-id'}})
+      #   Note this does not update any fields on the customer.
+      #
+      # To use a customer, but link a new card to the customer:
+      #     purchase(money, card_nonce, {customer: {id: 'customer-id'}})
+      #   Note this does not update any other fields on the customer.
+      #
+      # As this may take multiple requests, it returns a MultiResponse.
+      def purchase(money, card_nonce, options={})
+        raise ArgumentError('money required') if money.nil?
+        if card_nonce.nil?
+          requires!(options, :customer)
+          requires!(options[:customer], :card_id, :id)
+        end
+        if card_nonce && options[:customer] && options[:customer][:card_id]
+          raise ArgumentError('Cannot call with both card_nonce and options[:customer][:card_id], choose one.')
+        end
+
+        post = {}.merge(options.slice(:buyer_email_address, :delay_capture, :note, :reference_id))
+        add_idempotency_key(post, options)
+        add_amount(post, money, options)
+        add_address(post, options)
+        post[:reference_id] = options[:order_id] if options[:order_id]
+        post[:note] = options[:description] if options[:description]
+
+        MultiResponse.run do |r|
+          if options[:customer] && card_nonce
+            # Since customer was passed in, create customer (if needed) and store card (always in here).
+            options[:customer][:customer_id] = options[:customer][:id] if options[:customer][:id] # To make store() happy.
+            r.process { store(card_nonce, options[:customer]) }
+
+            # If we just created a customer.
+            if options[:customer][:id].nil?
+              options[:customer][:id] = r.responses.first.params['customer']['id']
```

Comment:
yes, they come back as json - this really bothered me though b/c options are all assumed to take symbols too. and since options can have nested maps, i didn't see other gateways use any deep conversion of strings acting as symbols or whatever that active support method was